### PR TITLE
Add attachments support to slack module.

### PR DESCRIPTION
**Pull request by nsg**
Currently attachments are only used to add a coloured line beside the text (when we use `color`):

``` python
if color == 'normal':
    payload = dict(text=text)
else:
    payload = dict(attachments=[dict(text=text, color=color)])
```

This PR adds the `attachments` keyword so you can use the full API (multiple attachments, fields, colours, images, ...). I have tried to keep the old behaviour.

``` python
payload = {}
if color == "normal" and text is not None:
     payload = dict(text=text)
elif text is not None:
    payload = dict(attachments=[dict(text=text, color=color)])
```

`msg` is `text` in the code above, if the `msg` field is used (was mandatory) the old behaviour is used.

Basically I have:
- Made the `msg` field optional
- Added `attachments` that contains a YAML representation of the JSON Attachments API found here: https://api.slack.com/docs/attachments
## Sample

``` yaml
    - name: Send slack message
      sudo: no
      local_action:
        module: slack
        domain: "{{ slack_domain }}"
        token: "{{ slack_token }}"
        channel: "{{ slack_channel }}"
        attachments:
          - text: "Display my system load on host A and B"
            color: "#ff00dd"
            title: "System load"
            fields:
              - title: "System A"
                value: "load average: 0,74, 0,66, 0,63"
                short: "true"
              - title: "System B"
                value: "load average: 5,16, 4,64, 2,43"
                short: "true"
```

Image of Yaktocat
## Note
- You can use both `msg` and `attachments`. It works and displays correctly in slack but I'm not sure if the format is really supported.
- You can write a empty message by omitting both `msg` and `attachments`, but I do not think that is a huge problem.
## Feedback

This is my first PR to Ansible so please tell me if everything is in order. I'm happy to improve/change the code, feedback is welcome!

~~This PR is **not** ready to be merged, I need to test it. It is here for you to test it and give feedback. I will try to test it properly in a day or two.~~

### CODE REVIEW

**Comment by nsg** [comment]
I have tested it both with `hacking/test-module` and for real (we are using this extension at work). Maybe the documentation needs to be rephrased? Love some input on that.

**Comment by gregdek** [comment]
@ramondelafuente Please review this PR; we will merge with +1 from you.

**Comment by gregdek** [comment]
@nsg Thanks for submitting this pull request. Unfortunately, it appears that the underlying code has changed since your submission, and this PR now needs to be rebased. 

If you believe that this PR still applies, please rebase and we will put it into our community review process as soon as possible.

**Comment by nsg** [comment]
@gregdek Done! It's now rebased on top of devel.

File: notification/slack.py [hunk 1]
```diff
@@ -24,7 +25,7 @@
 description:
     - The M(slack) module sends notifications to U(http://slack.com) via the Incoming WebHook integration
 version_added: 1.6
-author: "Ramon de la Fuente (@ramondelafuente)"
+author: Ramon de la Fuente <ramon@delafuente.nl>, Stefan Berggren <nsg@nsg.cc>
```
Comment (nsg) [inline notification/slack.py]: Is this appropriate? Should I remove my name? Is this reserved to the module maintainer or something else maybe? Same question for [L4](https://github.com/nsg/ansible-modules-extras/blob/full-attachments-support/notification/slack.py#L4)

I noticed that there are plenty of people that haven’t added their names.

Comment (bcoca) [inline notification/slack.py]: we are looking at changing it as 'author' is a misnomer at this point, it really is 'module point persons' or 'maintainers'

**Comment by devth** [comment]
:+1:

File: notification/slack.py (hunk 1 above)
Comment (bcoca) [inline notification/slack.py]: also, author should be a list when more than one:

```
author: 
   - Ramon de la Fuente <ramon@delafuente.nl>
   - Stefan Berggren <nsg@nsg.cc>
```

Comment (nsg) [inline notification/slack.py]: @bcoca I see, thank for the information. I think I will do the easy thing, just remove my name :)

**Comment by nsg** [comment]
I did a small update where I removed my name from `author` (see conversation on the outdated diff), I also found a few misplaced tabs in the documentation.

**Comment by gregdek** [comment]
Thanks @nsg -- sorry we didn't see that you'd rebased earlier. Still working out workflow kinks.

@ramondelafuente Please review this PR to make sure it adheres to the following guidelines:

http://docs.ansible.com/developing_modules.html#module-checklist

If it passes these guidelines, and if you believe it’s a good PR otherwise, please add a comment with "shipit" in the text, and we will flag it for inclusion.

If it fails these guidelines, please add a comment with "needs_revision" in the text, along with the needed changes, and we will flag it for revision.

**Comment by robynbergeron** [comment]
@ramondelafuente -- ping? :) Just need your approval to ship this. :)

**Comment by ramondelafuente** [comment]
It looks good... It's just that I was going to run a test to see if it did what I expect, but the clock hasn't been agreeing with my plan.

**Comment by ramondelafuente** [comment]
Also, the older code isn't in line with the newer guidelines, and it seems unfair to put that on someone adding functionality.
Example: not all `required: false` parameters have `default` values documented.

**Comment by ramondelafuente** [comment]
needs_revision

**Documentation**
- not all `required:false` parameters in the docs have defaults
- the `token` parameter couls be considered a secret argument and should have `no_log=True`

**Code**
- does not call `main` in a conditional

**Comment by ramondelafuente** [comment]
And finally: @nsg 
Sorry for the "needs_review" tag, it's just that my own work in the past doesn't follow the current guidelines. Thanks a lot for the addition of the full attachments API, nice work :-)

**Comment by robynbergeron** [comment]
Hi  @nsg -- thanks for submitting this PR and being patient :)

@ramondelafuente -- thanks for your comments. I assume the needs_revision indicates that you are going to add the details needed to bring this up to current guidelines (since you mentioned the "seems unfair" thing)? 

Just want to confirm who should be taking action here, so we can keep this moving. 

Thanks so much! :D

**Comment by nsg** [comment]
@robynbergeron No problem, and thank you for the messages :)

@ramondelafuente If you feel that your time is limited I can probably add the missing things this weekend, just one problem, I'm a little unsure what I need to do.

**Comment by gregdek** [comment]
@nsg What @ramondelafuente is saying is that the module is not in accordance to the module guidelines here (which have actually changed since he wrote the original:)

http://docs.ansible.com/developing_modules.html#module-checklist

It's all pretty simple stuff, so a few simple revisions and we should be good to go. Thanks!

**Comment by ramondelafuente** [comment]
@nsg I have submitted a PR to your branch with the additional work. As soon as you merge this PR will reflect the changes (and then I think I should add a label here that it's good to go...)

**Comment by nsg** [comment]
@ramondelafuente Merged and tested with the latest code from HEAD. Looks good.
@gregdek Thank you for the link and clarification.

**Comment by robynbergeron** [comment]
@nsg -- thanks for the updates. 

@ramondelafuente -- looks like you can add that label now about it being good to go, but would love to double-check that everything you were expecting to be merged was merged and that it's really good to go, before I mark it for more merging. :)

Thanks, friends!

**Comment by ramondelafuente** [comment]
Wait for it...
...
wait for it...
...

Shipit!

**Pull request merged**